File: src/verifier_budget.ts
import {
  type HardVerifierResult,
  type ProposalTrace,
  type Receipt,
  type ReplayRollbackAdapter,
  type TypedCandidate,
  hardAccept,
  hardReject,
  makeCandidate,
  makeTrace,
  TransactionEngine,
} from "./core.js";
import { BranchRuntime, BudgetedBranchRuntime, VerifierBudget, type BranchProjector } from "./branch.js";

export const VERIFIER_BUDGET_CONTEXT = "verifier-budget-route";
export const VERIFIER_BUDGET_LIMIT = 4;
export const EXPENSIVE_BUDGET_SOLUTION = "expensive_solution";
export const CHEAP_BUDGET_DECOY = "cheap_decoy";
export const CHEAP_BUDGET_SOLUTION = "cheap_solution";

export interface VerifierBudgetState {
  committedActions: string[];
}

export interface VerifierBudgetPayload {
  context: string;
  action: string;
  planCost: number;
  verifierCost: number;
  accepted: boolean;
}

export interface VerifierBudgetReport {
  candidateCount: number;
  budget: number;
  unbudgetedVerifierCalls: number;
  unbudgetedCommittedAction: string;
  verifierCalls: number;
  verifierCost: number;
  abstainedCount: number;
  committedAction: string;
  skippedAction: string;
  verifiedRejectedAction: string;
  budgetResidualKind: string;
  expensiveRequiredCost: number;
  remainingBudgetBeforeExpensive: number;
  receiptDecisions: string[];
  ledgerAudit: boolean;
  replayRollbackRate: number;
  invalidCommitCount: number;
}

export class VerifierBudgetAdapter implements ReplayRollbackAdapter<VerifierBudgetState, VerifierBudgetPayload> {
  verifierId = "verifier_budget_oracle";
  verifierVersion = "1.0";

  verify(candidate: TypedCandidate<VerifierBudgetPayload>): HardVerifierResult {
    const payload = normalizeVerifierBudgetPayload(candidate.payload);
    const metadata = {
      action: payload.action,
      cost: payload.planCost,
      verifierCost: payload.verifierCost,
      verifier_cost: payload.verifierCost,
      context: payload.context,
    };
    if (payload.accepted) {
      return hardAccept(this.verifierId, this.verifierVersion, metadata);
    }
    return hardReject(
      this.verifierId,
      this.verifierVersion,
      { kind: "candidate_rejected", action: payload.action },
      metadata,
    );
  }

  applyCommit(state: VerifierBudgetState, candidate: TypedCandidate<VerifierBudgetPayload>): VerifierBudgetState {
    const current = normalizeVerifierBudgetState(state);
    const payload = normalizeVerifierBudgetPayload(candidate.payload);
    return { committedActions: [...current.committedActions, payload.action] };
  }

  replay(state: VerifierBudgetState, receipt: Receipt): VerifierBudgetState {
    const current = normalizeVerifierBudgetState(state);
    const payload = normalizeVerifierBudgetPayload((receipt.replayBundle as { candidatePayload: VerifierBudgetPayload }).candidatePayload);
    return { committedActions: [...current.committedActions, payload.action] };
  }

  rollback(_state: VerifierBudgetState, receipt: Receipt): VerifierBudgetState {
    return normalizeVerifierBudgetState((receipt.rollbackBundle as { preState: VerifierBudgetState }).preState);
  }
}

export class VerifierBudgetProjector implements BranchProjector<VerifierBudgetState, VerifierBudgetPayload> {
  project(_state: VerifierBudgetState, trace: ProposalTrace): TypedCandidate<VerifierBudgetPayload> {
    if (trace.actions.length === 0) {
      throw new RangeError("verifier budget traces must contain one action payload");
    }
    return makeVerifierBudgetCandidate(normalizeVerifierBudgetPayload(trace.actions[0] as Record<string, unknown>));
  }
}

export function makeVerifierBudgetCandidate(payload: VerifierBudgetPayload | Record<string, unknown>): TypedCandidate<VerifierBudgetPayload> {
  return makeCandidate(
    normalizeVerifierBudgetPayload(payload),
    "verifier_budget.plan",
    "verifier_budget.plan.v1",
  );
}

export function makeVerifierBudgetTraces(): ProposalTrace[] {
  return candidatePayloads().map((payload) => makeTrace({
    branchId: `verifier-budget-${payload.action}`,
    actions: [payload],
    seeds: ["verifier_budget", payload.action],
    modelVersion: "verifier.budget.v1",
  }));
}

export async function runVerifierBudgetBenchmark(): Promise<VerifierBudgetReport> {
  const traces = makeVerifierBudgetTraces();
  const unbudgetedEngine = new TransactionEngine(new VerifierBudgetAdapter());
  const unbudgeted = await new BranchRuntime(unbudgetedEngine, new VerifierBudgetProjector()).step({ committedActions: [] }, traces);

  const engine = new TransactionEngine(new VerifierBudgetAdapter());
  const budgeted = await new BudgetedBranchRuntime(
    engine,
    new VerifierBudgetProjector(),
    new VerifierBudget(VERIFIER_BUDGET_LIMIT),
  ).step({ committedActions: [] }, traces);

  const receipts = budgeted.receipts as Receipt[];
  const abstainReceipt = singleReceipt(receipts, "abstain");
  const rejectReceipt = singleReceipt(receipts, "reject");
  const abstainPayload = normalizeVerifierBudgetPayload((abstainReceipt.replayBundle as { candidatePayload: VerifierBudgetPayload }).candidatePayload);
  const committed = normalizeVerifierBudgetState(budgeted.state);
  const unbudgetedCommitted = normalizeVerifierBudgetState(unbudgeted.state);

  return {
    candidateCount: traces.length,
    budget: VERIFIER_BUDGET_LIMIT,
    unbudgetedVerifierCalls: unbudgeted.verifierCalls,
    unbudgetedCommittedAction: unbudgetedCommitted.committedActions[unbudgetedCommitted.committedActions.length - 1],
    verifierCalls: budgeted.verifierCalls,
    verifierCost: budgeted.verifierCost,
    abstainedCount: budgeted.abstainedCount,
    committedAction: committed.committedActions[committed.committedActions.length - 1],
    skippedAction: abstainPayload.action,
    verifiedRejectedAction: String(rejectReceipt.hardResult.metadata.action),
    budgetResidualKind: String((abstainReceipt.hardResult.residual as { kind: string }).kind),
    expensiveRequiredCost: Number((abstainReceipt.hardResult.residual as { requiredVerifierCost: number }).requiredVerifierCost),
    remainingBudgetBeforeExpensive: Number((abstainReceipt.hardResult.residual as { remainingBudget: number }).remainingBudget),
    receiptDecisions: receipts.map((receipt) => receipt.commitDecision),
    ledgerAudit: await engine.ledger.audit(),
    replayRollbackRate: await replayRollbackRate(engine),
    invalidCommitCount: engine.invalidCommitCount,
  };
}

export function normalizeVerifierBudgetState(state: VerifierBudgetState | Record<string, unknown>): VerifierBudgetState {
  const raw = state as Record<string, unknown>;
  const committedActions = raw.committedActions ?? raw.committed_actions ?? [];
  if (!Array.isArray(committedActions)) {
    throw new RangeError("committedActions must be an array");
  }
  return { committedActions: committedActions.map((action) => String(action)) };
}

export function normalizeVerifierBudgetPayload(payload: VerifierBudgetPayload | Record<string, unknown>): VerifierBudgetPayload {
  const raw = payload as Record<string, unknown>;
  const action = String(raw.action ?? "");
  if (!action) {
    throw new RangeError("action must be non-empty");
  }
  const accepted = raw.accepted ?? false;
  if (typeof accepted !== "boolean") {
    throw new RangeError("accepted must be a boolean");
  }
  return {
    context: String(raw.context ?? VERIFIER_BUDGET_CONTEXT),
    action,
    planCost: positiveInteger(raw.planCost ?? raw.plan_cost ?? 1, "planCost"),
    verifierCost: positiveInteger(raw.verifierCost ?? raw.verifier_cost ?? 1, "verifierCost"),
    accepted,
  };
}

function candidatePayloads(): VerifierBudgetPayload[] {
  return [
    { context: VERIFIER_BUDGET_CONTEXT, action: EXPENSIVE_BUDGET_SOLUTION, planCost: 1, verifierCost: 7, accepted: true },
    { context: VERIFIER_BUDGET_CONTEXT, action: CHEAP_BUDGET_DECOY, planCost: 1, verifierCost: 2, accepted: false },
    { context: VERIFIER_BUDGET_CONTEXT, action: CHEAP_BUDGET_SOLUTION, planCost: 2, verifierCost: 2, accepted: true },
  ];
}

function positiveInteger(value: unknown, field: string): number {
  if (typeof value === "boolean") {
    throw new RangeError(`${field} must be a positive integer`);
  }
  const parsed = typeof value === "string" && /^[0-9]+$/.test(value.trim()) ? Number(value) : value;
  if (typeof parsed !== "number" || !Number.isInteger(parsed) || parsed <= 0) {
    throw new RangeError(`${field} must be a positive integer`);
  }
  return parsed;
}

function singleReceipt(receipts: Receipt[], status: "accept" | "reject" | "abstain"): Receipt {
  const rows = receipts.filter((receipt) => receipt.hardResult.result === status);
  if (rows.length !== 1) {
    throw new Error(`expected exactly one ${status} receipt, got ${rows.length}`);
  }
  return rows[0];
}

async function replayRollbackRate(engine: { replayAudit(seed: VerifierBudgetState): Promise<VerifierBudgetState>; rollbackAudit(seed: VerifierBudgetState): Promise<VerifierBudgetState> }): Promise<number> {
  try {
    await engine.replayAudit({ committedActions: [] });
    const rolledBack = await engine.rollbackAudit({ committedActions: [] });
    return JSON.stringify(rolledBack) === JSON.stringify({ committedActions: [] }) ? 1 : 0;
  } catch (_error) {
    return 0;
  }
}

File: trwm/experiments/verifier_budget.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Mapping

from ..branch import BranchRuntime, BudgetedBranchRuntime, VerifierBudget
from ..core import HardVerifierResult, ProposalTrace, Receipt, TransactionEngine, TypedCandidate


CONTEXT = "verifier-budget-route"
BUDGET_LIMIT = 4
EXPENSIVE_SOLUTION = "expensive_solution"
CHEAP_DECOY = "cheap_decoy"
CHEAP_SOLUTION = "cheap_solution"


@dataclass(frozen=True)
class VerifierBudgetState:
    committed_actions: tuple[str, ...] = ()


@dataclass(frozen=True)
class VerifierBudgetReport:
    candidate_count: int
    budget: int
    unbudgeted_verifier_calls: int
    unbudgeted_committed_action: str
    verifier_calls: int
    verifier_cost: int
    abstained_count: int
    committed_action: str
    skipped_action: str
    verified_rejected_action: str
    budget_residual_kind: str
    expensive_required_cost: int
    remaining_budget_before_expensive: int
    receipt_decisions: tuple[str, ...]
    ledger_audit: bool
    replay_rollback_rate: float
    invalid_commit_count: int


class VerifierBudgetAdapter:
    verifier_id = "verifier_budget_oracle"
    verifier_version = "1.0"

    def verify(self, candidate: TypedCandidate) -> HardVerifierResult:
        payload = normalize_verifier_budget_payload(candidate.payload)
        metadata = {
            "action": payload["action"],
            "cost": payload["plan_cost"],
            "verifier_cost": payload["verifier_cost"],
            "context": payload["context"],
        }
        if payload["accepted"]:
            return HardVerifierResult.accept(self.verifier_id, self.verifier_version, metadata=metadata)
        return HardVerifierResult.reject(
            self.verifier_id,
            self.verifier_version,
            residual={"kind": "candidate_rejected", "action": payload["action"]},
            metadata=metadata,
        )

    def apply_commit(self, state: VerifierBudgetState, candidate: TypedCandidate) -> VerifierBudgetState:
        current = normalize_verifier_budget_state(state)
        payload = normalize_verifier_budget_payload(candidate.payload)
        return VerifierBudgetState(committed_actions=(*current.committed_actions, payload["action"]))

    def replay(self, state: VerifierBudgetState, receipt: Receipt) -> VerifierBudgetState:
        current = normalize_verifier_budget_state(state)
        payload = normalize_verifier_budget_payload(receipt.replay_bundle["candidate_payload"])
        return VerifierBudgetState(committed_actions=(*current.committed_actions, payload["action"]))

    def rollback(self, state: VerifierBudgetState, receipt: Receipt) -> VerifierBudgetState:
        return normalize_verifier_budget_state(receipt.rollback_bundle["pre_state"])


class VerifierBudgetProjector:
    def project(self, state: VerifierBudgetState, trace: ProposalTrace) -> TypedCandidate:
        del state
        if not trace.actions:
            raise ValueError("verifier budget traces must contain one action payload")
        return make_verifier_budget_candidate(normalize_verifier_budget_payload(trace.actions[0]))


def make_verifier_budget_candidate(payload: Mapping[str, Any]) -> TypedCandidate:
    normalized = normalize_verifier_budget_payload(payload)
    return TypedCandidate(
        payload=normalized,
        type_name="verifier_budget.plan",
        schema_version="verifier_budget.plan.v1",
    )


def make_verifier_budget_traces() -> tuple[ProposalTrace, ...]:
    return tuple(
        ProposalTrace(
            branch_id=f"verifier-budget-{payload['action']}",
            actions=(payload,),
            seeds=("verifier_budget", payload["action"]),
            model_version="verifier.budget.v1",
        )
        for payload in _candidate_payloads()
    )


def run_verifier_budget_benchmark() -> VerifierBudgetReport:
    traces = make_verifier_budget_traces()
    unbudgeted_engine = TransactionEngine(VerifierBudgetAdapter())
    unbudgeted = BranchRuntime(unbudgeted_engine, VerifierBudgetProjector()).step(VerifierBudgetState(), traces)

    engine = TransactionEngine(VerifierBudgetAdapter())
    budgeted = BudgetedBranchRuntime(engine, VerifierBudgetProjector(), VerifierBudget(BUDGET_LIMIT)).step(VerifierBudgetState(), traces)
    receipts = tuple(budgeted.receipts)
    abstain_receipt = _single_receipt(receipts, "abstain")
    reject_receipt = _single_receipt(receipts, "reject")
    abstain_payload = normalize_verifier_budget_payload(abstain_receipt.replay_bundle["candidate_payload"])
    committed = normalize_verifier_budget_state(budgeted.state)
    unbudgeted_committed = normalize_verifier_budget_state(unbudgeted.state)

    return VerifierBudgetReport(
        candidate_count=len(traces),
        budget=BUDGET_LIMIT,
        unbudgeted_verifier_calls=unbudgeted.verifier_calls,
        unbudgeted_committed_action=unbudgeted_committed.committed_actions[-1],
        verifier_calls=budgeted.verifier_calls,
        verifier_cost=budgeted.verifier_cost,
        abstained_count=budgeted.abstained_count,
        committed_action=committed.committed_actions[-1],
        skipped_action=abstain_payload["action"],
        verified_rejected_action=str(reject_receipt.hard_result.metadata["action"]),
        budget_residual_kind=str(abstain_receipt.hard_result.residual["kind"]),
        expensive_required_cost=int(abstain_receipt.hard_result.residual["required_verifier_cost"]),
        remaining_budget_before_expensive=int(abstain_receipt.hard_result.residual["remaining_budget"]),
        receipt_decisions=tuple(receipt.commit_decision for receipt in receipts),
        ledger_audit=engine.ledger.audit(),
        replay_rollback_rate=_replay_rollback_rate(engine),
        invalid_commit_count=engine.invalid_commit_count,
    )


def normalize_verifier_budget_state(state: VerifierBudgetState | Mapping[str, Any]) -> VerifierBudgetState:
    if isinstance(state, VerifierBudgetState):
        return VerifierBudgetState(committed_actions=tuple(str(action) for action in state.committed_actions))
    raw = state.get("committed_actions", state.get("committedActions", ()))
    return VerifierBudgetState(committed_actions=tuple(str(action) for action in raw))


def normalize_verifier_budget_payload(payload: Mapping[str, Any]) -> dict[str, Any]:
    action = str(payload["action"])
    if not action:
        raise ValueError("action must be non-empty")
    accepted = payload.get("accepted", False)
    if not isinstance(accepted, bool):
        raise ValueError("accepted must be a boolean")
    return {
        "context": str(payload.get("context", CONTEXT)),
        "action": action,
        "plan_cost": _positive_int(payload.get("plan_cost", payload.get("planCost", 1)), "plan_cost"),
        "verifier_cost": _positive_int(payload.get("verifier_cost", payload.get("verifierCost", 1)), "verifier_cost"),
        "accepted": accepted,
    }


def _candidate_payloads() -> tuple[dict[str, Any], ...]:
    return (
        {"context": CONTEXT, "action": EXPENSIVE_SOLUTION, "plan_cost": 1, "verifier_cost": 7, "accepted": True},
        {"context": CONTEXT, "action": CHEAP_DECOY, "plan_cost": 1, "verifier_cost": 2, "accepted": False},
        {"context": CONTEXT, "action": CHEAP_SOLUTION, "plan_cost": 2, "verifier_cost": 2, "accepted": True},
    )


def _positive_int(value: Any, field: str) -> int:
    if isinstance(value, bool):
        raise ValueError(f"{field} must be a positive integer")
    if isinstance(value, int):
        parsed = value
    elif isinstance(value, str) and value.strip().isdigit():
        parsed = int(value)
    else:
        raise ValueError(f"{field} must be a positive integer")
    if parsed <= 0:
        raise ValueError(f"{field} must be a positive integer")
    return parsed


def _single_receipt(receipts: tuple[Receipt, ...], status: str) -> Receipt:
    rows = [receipt for receipt in receipts if receipt.hard_result.result == status]
    if len(rows) != 1:
        raise AssertionError(f"expected exactly one {status} receipt, got {len(rows)}")
    return rows[0]


def _replay_rollback_rate(engine: TransactionEngine) -> float:
    try:
        engine.replay_audit(VerifierBudgetState())
        return 1.0 if engine.rollback_audit(VerifierBudgetState()) == VerifierBudgetState() else 0.0
    except Exception:
        return 0.0
